results-processor/processor.py

```diff
@@ -18,200 +21,282 @@
-
-    Datastore exceptions may be raised.
-
-    Returns:
-        A tuple (username, password).
-    """
-    global _auth
-    if _auth is None:
-        user = _datastore.get(_datastore.key('Uploader', '_processor'))
-        _auth = (user['Username'], user['Password'])
-    return _auth
-
-
-def _find_run_by_raw_results(raw_results_url):
-    """Returns true if an existing run already has the same raw_results_url."""
-    q = _datastore.query(kind='TestRun')
-    q.add_filter('RawResultsURL', '=', raw_results_url)
-    q.keys_only()
-    run = list(q.fetch(limit=1))
-    return len(run) > 0
+class Processor(object):
+    USERNAME = '_processor'
+
+    def __init__(self):
+        self._datastore = datastore.Client()
+        # Temporary directories to be created in __enter__:
+        self._temp_dir = '/tempdir/for/raw/results/screenshots'
+        self._upload_dir = '/tempdir/for/split/results'
+
+        # A tuple of (username, password) for processor.
+        self.auth = self._get_auth()
+        # Local paths to downloaded results and screenshots:
+        self.results = []
+        self.screenshots = []
+        # To be loaded/initialized later:
+        self.report = wptreport.WPTReport()
+        self.test_run_id = ''
+
+    def __enter__(self):
+        self._temp_dir = tempfile.mkdtemp()
+        self._upload_dir = tempfile.mkdtemp()
+        return self
+
+    def __exit__(self, *args):
+        shutil.rmtree(self._temp_dir)
+        shutil.rmtree(self._upload_dir)
+
+    def _get_auth(self):
+        user = self._datastore.get(
+            self._datastore.key('Uploader', self.USERNAME))
+        return (user['Username'], user['Password'])
+
+    @property
+    def raw_results_gs_url(self):
+        return 'gs://{}/{}/report.json'.format(
+            config.raw_results_bucket(), self.report.sha_product_path)
+
+    @property
+    def raw_results_url(self):
+        return gsutil.gs_to_public_url(self.raw_results_gs_url)
+
+    @property
+    def results_gs_url(self):
+        return 'gs://{}/{}'.format(
+            config.results_bucket(), self.report.sha_summary_path)
+
+    @property
+    def results_url(self):
+        return gsutil.gs_to_public_url(self.results_gs_url),
+
+    def check_existing_run(self):
+        """Returns true if an existing run already has raw_results_url.
+
+        This is used to abort early if the result already exists in Datastore.
+        It is safe because raw_results_url contains both the full revision &
+        checksum of the report content, unique enough to use as a UID.
+
+        Datastore does not support a query-and-put transaction, so this is
+        only a best effort to avoid duplicate runs.
+        """
+        q = self._datastore.query(kind='TestRun')
+        q.add_filter('RawResultsURL', '=', self.raw_results_url)
+        q.keys_only()
+        run = list(q.fetch(limit=1))
+        return len(run) > 0
+
+    @staticmethod
+    def known_extension(path):
```

Comment:
Looks like this is used to give temporary files the right extension. But does that matter, can't they be called anything as long as they have the right content?

(This is in the first refactoring commit, but seems like it's new?)

Reply:
Yes. And it matters because the loaders (both for screenshots and for reports) rely on file extensions instead of magic bytes to detect gzip. This is borderline "refactor": since the API requires uploaded files to be gzipped, the non-gzipped code path won't be exercised in reality. And the exact refactoring would be always returning ".gz" here. Sorry for making the already big change more confusing.

Reply:
Nah, it was nicely split, just this bit caught my eye :)